From 32a79b68b79b32ee97e3931e0a94bc33767b1f49 Mon Sep 17 00:00:00 2001
From: Arman Uguray <armansito@chromium.org>
Date: Fri, 20 Feb 2015 17:56:47 -0800
Subject: [PATCH 1/3] gdbus: Don't refresh objects/props if disconnected

If g_dbus_client_set_proxy_handlers gets called from within a
proxy_removed callback, the code may end up refreshing the proxy's
properties and incorrectly access the client's proxy_list as it gets
freed. This patch fixes this, so that get_managed_objects does nothing
if it gets called during a service disconnect.
---
 gdbus/client.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/gdbus/client.c b/gdbus/client.c
index eb68a0f..238b348 100644
--- a/gdbus/client.c
+++ b/gdbus/client.c
@@ -1107,6 +1107,9 @@ static void get_managed_objects(GDBusClient *client)
 {
 	DBusMessage *msg;
 
+	if (!client->connected)
+		return;
+
 	if (!client->proxy_added && !client->proxy_removed) {
 		refresh_properties(client);
 		return;
@@ -1142,13 +1145,13 @@ static void service_connect(DBusConnection *conn, void *user_data)
 
 	g_dbus_client_ref(client);
 
+	client->connected = TRUE;
+
 	if (client->connect_func)
 		client->connect_func(conn, client->connect_data);
 
 	get_managed_objects(client);
 
-	client->connected = TRUE;
-
 	g_dbus_client_unref(client);
 }
 
@@ -1156,13 +1159,13 @@ static void service_disconnect(DBusConnection *conn, void *user_data)
 {
 	GDBusClient *client = user_data;
 
+	client->connected = FALSE;
+
 	g_list_free_full(client->proxy_list, proxy_free);
 	client->proxy_list = NULL;
 
-	if (client->disconn_func) {
+	if (client->disconn_func)
 		client->disconn_func(conn, client->disconn_data);
-		client->connected = FALSE;
-	}
 }
 
 static DBusHandlerResult message_filter(DBusConnection *connection,
-- 
1.9.1

